src/commands/publish.rs

```diff
@@ -140,13 +117,53 @@ pub fn publish(
     } else {
         let upload_client = http::legacy_auth_client(user);
 
+        deploy::pre_upload(user, target, &deployments)?;
         upload::script(&upload_client, &target, None)?;
         deploy(target)?;
     }
 
     Ok(())
 }
 
+fn build_output_message(deploy_results: deploy::DeployResults, target_name: String, out: Output) {
+    let deploy::DeployResults {
+        urls,
+        schedules,
+        durable_object_namespaces,
+    } = deploy_results;
+
+    let mut fragments: Vec<String> = vec![];
+    fragments.push(String::from("Successfully published your script"));
+
+    for entry in [
+        ("to \n", &urls),
+        ("with this schedule\n", &schedules),
+        (
+            "implementing these durable object namespaces\n",
+            &durable_object_namespaces,
+        ),
+    ]
+    .iter()
+    {
+        if !entry.1.is_empty() {
+            fragments.push(format!("{} {}", entry.0, entry.1.join("\n ")))
+        }
+    }
+
+    let result_msg = fragments.join(" ");
```

Comment:
I'm having a hard time reading this, but I have an idea about what it's doing... Perhaps it could be rewritten to make it more obvious from a quick glance what exactly it's doing?

```rust
let deploy::DeployResults {
    urls,
    schedules,
    durable_object_namespaces: namespaces,
} = deploy_results;

let msg = "Successfully published your script".to_owned();
if !urls.is_empty() {
    msg.push_str(&format!("to\n {}\n", urls.join("\n ")));
}
if !schedules.is_empty() {
    msg.push_str(&format!("with this schedule\n {}\n", schedules.join("\n ")));
}
if !namespaces.is_empty() {
    msg.push_str(&format!("implementing these durable object namespaces\n {}\n", namespaces.join("\n ")));
}
```

Reply:
Actually you could make it even better with `write!`:

```rust
let deploy::DeployResults {
    urls,
    schedules,
    durable_object_namespaces: namespaces,
} = deploy_results;

let mut msg = "Successfully published your script".to_owned();
if !urls.is_empty() {
    write!(&mut msg, "to\n {}\n", urls.join("\n ")).unwrap();
}
if !schedules.is_empty() {
    write!(&mut msg, "with this schedule\n {}\n", schedules.join("\n "));
}
if !namespaces.is_empty() {
    write!(&mut msg, "implementing these durable object namespaces\n {}\n", namespaces.join("\n "));
}
```

Reply:
yes, this is much better :)

Reply:
gonna go with the first one since write returns some weird result type and I don't like bare unwraps

Reply:
It's just `std::fmt::Result`. `format!` does the exact same thing but it's hidden behind the function it calls
```rust
pub fn format(args: Arguments<'_>) -> string::String {
    let capacity = args.estimated_capacity();
    let mut output = string::String::with_capacity(capacity);
    output.write_fmt(args).expect("a formatting trait implementation returned an error");
    output
}
```

Reply:
The point about bare unwraps isn't that panics never happen. Format documents that it panics. The original implementation also panics in "shouldn't happen" cases, e.g. vec push fails because memory can't be alllocated.

The point is that historical evidence shows that bare unwraps we wrote caused confusing errors for wrangler users.

We should get out of the habit of adding more bare unwrap and think about the error case.  If it's really not reasonable to handle, at least use expect with a message that's easier for users to understand and easier for us to track down. 

In this case just using format seems better to me.